gadget/update.go

```diff
@@ -326,6 +327,111 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
+		AssumeCreatablePartitionsCreated: true,
+	}
+	if err := EnsureLayoutCompatibility(expLayout, diskLayout, opts); err != nil {
+		return res, fmt.Errorf("volume %s is not compatible with disk %s: %v", vol.Name, dev, err)
+	}
+
+	// also get a Disk{} interface for this device
+	disk, err := disks.DiskFromDeviceName(dev)
+	if err != nil {
+		return res, fmt.Errorf("cannot get disk for device %s: %v", dev, err)
+	}
+
+	mappedStructures := make([]DiskStructureDeviceTraits, 0, len(diskLayout.Structure))
+
+	diskPartitions, err := disk.Partitions()
+	if err != nil {
+		return res, fmt.Errorf("cannot get partitions for disk device %s: %v", dev, err)
+	}
+
+	// make an easy map of start offsets to to partitions for lookup
+	diskPartitionsByOffset := map[uint64]disks.Partition{}
+	for _, p := range diskPartitions {
+		diskPartitionsByOffset[p.StartInBytes] = p
+	}
+
+	// create the traits for each structure looping over the laid out structure
+	// to ensure that extra partitions don't sneak in - we double check things
+	// again below this loop
+	for _, structure := range expLayout.LaidOutStructure {
+		// don't create traits for non-partitions, there is nothing we can
+		// measure on the disk about bare structures other than perhaps reading
+		// their content - the fact that bare structures do not overlap with
+		// real partitions will have been validated when the YAML was validated
+		// previously
+		if !structure.IsPartition() {
+			continue
+		}
+
+		part, ok := diskPartitionsByOffset[uint64(structure.StartOffset)]
+		if !ok {
+			// unexpected error - somehow this structure's start offset is not
+			// present in the OnDiskVolume, which is unexpected because we
+			// validated that the laid out volume structure matches the on disk
+			// volume
+			return res, fmt.Errorf("internal error: inconsistent disk structures from LaidOutVolume and disks.Disk: structure starting at %d missing on disk", structure.StartOffset)
+		}
+		ms := DiskStructureDeviceTraits{
+			Size:               quantity.Size(part.SizeInBytes),
+			Offset:             quantity.Offset(part.StartInBytes),
+			PartitionUUID:      part.PartitionUUID,
+			OriginalKernelPath: part.KernelDeviceNode,
+			OriginalDevicePath: part.KernelDevicePath,
+			PartitionType:      part.PartitionType,
+			PartitionLabel:     part.PartitionLabel,  // this will be empty on dos disks
+			FilesystemLabel:    part.FilesystemLabel, // blkid encoded
+			FilesystemUUID:     part.FilesystemUUID,  // blkid encoded
+			FilesystemType:     part.FilesystemType,
+		}
+
+		mappedStructures = append(mappedStructures, ms)
+
+		// delete this partition from the map
+		delete(diskPartitionsByOffset, uint64(structure.StartOffset))
+	}
+
+	// check that we don't have any extra partitions on the disk that are not
+	// accounted for in the layout - again this is unexpected since we confirmed
+	// above that the laid out volume was compatible with the on disk volume
+	if len(diskPartitions) != len(mappedStructures) {
```

Comment:
This check assumes full knowledge of how the `disk.Partitions()` works and how the `expLayout` is made; I need to go through the code and see where these variables are defined and assigned to, in order to understand what we are checking here. Unless there's a strong reason for using exactly these two variables in the condition, I'd rather change the condition to be:
```suggestion
	// we removed the partitions from the map as we were encountering them, so this should now be empty
	if len(diskPartitionsByOffset) != 0 {
```

Reply:
I re-worded this whole section when I had to refactor this to take into account the implicit system-data role case for UC16/UC18